pkg/apis/core/validation/validation.go

```diff
@@ -4043,6 +4043,28 @@ func ValidateService(service *core.Service) field.ErrorList {
 		ports[key] = true
 	}
 
+	// Validate TopologyKeys
+	if utilfeature.DefaultFeatureGate.Enabled(features.TopologyAwareServiceRouting) && len(service.Spec.TopologyKeys) > 0 {
```

Comment:
So, if we validate only if the feature gate is enabled, but the field is always there, then we can end up with services with invalid values stored in the api server, which may cause issues after the feature flag is turned on? How is this normally handled? @thockin

Reply:
According to @thockin there needs to be a function in pkg/registry/core/service/strategy.go that will nil this field if the feature gate is not enabled.

Reply:
Yes, we can set `TopologyKeys` to nil in `dropServiceDisabledFields()` if he feature gate is not enabled, but this was not merged yet: https://github.com/kubernetes/kubernetes/pull/79386/files#diff-eb2346d83fd23b4e48ff8fd239a97aef

How about we just cancel the feature gate check at first, just check the len() of `TopologyKeys`, we will check and store it into apiserver if  the feature gate is not enabled but `TopologyKeys` is not nil and len() is valid. And when the PR metioned above above get merged, we add this check to `dropServiceDisabledFields()`:
``` go
	// Drop TopologyKeys if TopologyAwareServiceRouting is not enabled
	if !utilfeature.DefaultFeatureGate.Enabled(features.TopologyAwareServiceRouting) && oldSvc != nil && oldSvc.Spec.TopologyKeys != nil {
		newSvc.Spec.TopologyKeys = nil
	}
```

Reply:
https://github.com/kubernetes/kubernetes/pull/81826

So you won't need to check the gate here.

Reply:
That's great